### PR TITLE
Promote IPv4/IPv6 dual-stack support to Stable in 1.23

### PR DESCRIPTION
**Pull request by bridgetkromhout**
We anticipate moving dual-stack to stable and removing the feature gate in 1.23.

Enhancements issue: https://github.com/kubernetes/enhancements/issues/563
Kubernetes PR: https://github.com/kubernetes/kubernetes/pull/104691

Signed-off-by: Bridget Kromhout <bridget@kromhout.org>

### CODE REVIEW

**Comment by tengqm** [comment]
/hold
This looks like a placeholder because the upstream change is not there yet.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:802]: Please leave this description in place. We don't remove feature descriptions when they graduate.

**Review by sftim — COMMENTED** [review]
Hi. We'll also need to revise some more pages to get the documentation ready for dual-stack as a stable feature:

- https://kubernetes.io/docs/concepts/cluster-administration/networking/
- https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/

and possibly:
- https://kubernetes.io/docs/concepts/services-networking/
- https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/dual-stack-support/

My view is that https://kubernetes.io/docs/concepts/services-networking/dual-stack/ (as a _concept_) should disappear, being superseded by a new _task_ page about configuring dual-stack networking. We should reword https://kubernetes.io/docs/concepts/services-networking/ so that it mentions dual-stack in the page text, and also link to the page about configuring it.

We might also want a page about making sure that your cluster runs single stack networking, as you won't be able to use a feature gate any more to achieve that.

**Comment by bridgetkromhout** [comment]
> My view is that https://kubernetes.io/docs/concepts/services-networking/dual-stack/ (as a concept) should disappear, being superseded by a new task page about configuring dual-stack networking.

We can definitely move many examples to the tasks page. We'll also want to keep a concept page, as there are details that may confuse people encountering this for the first time. I'll submit an update to this PR.

**Comment by sftim** [comment]
I'm imagining that at some point when you visit https://kubernetes.io/docs/concepts/services-networking/dual-stack/ you are redirected to https://kubernetes.io/docs/concepts/services-networking#dual-stack

I think we should then have pages in the reference section about networking that cover the 3 cases we support:
- dual stack
- IPv4-only
- IPv6-only

and those can go into lots of detail. The reference section is the right home for that detail. Treating each of those 3 cases as equally valid feels important; people might well prefer to run a single-stack cluster and we shouldn't make them feel like second class readers; the same is true for people who want to run dual-stack.

I'm happy to help tidying up this effort; I've got general IPv6 experience dating back well over a decade. I've got pretty much zero experience of IPv6 with Kubernetes though.

**Comment by sftim** [comment]
@kubernetes/sig-network-misc this looks fine to me in terms of there being no problematic omissions or inaccuracies - would someone from the SIG be willing to add a tech sign off to say you agree?

(As far as I'm concerned, we can get that signoff in place even if the hold remains).

**Comment by bridgetkromhout** [comment]
I was first writing a release blog, and now I'm getting back to this PR. Updates will come in short order.

**Comment by bridgetkromhout** [comment]
While @sftim has excellent ideas for docs refactoring, those may not fit in the desired timeframe - while this PR is an attempt to make the docs for 1.23 as accurate as is feasible.

**Review by khenidak — APPROVED** [review]

**Comment by bridgetkromhout** [comment]
Hi @jlbutler @jimangel @nate-double-u and all of the docs team - while we can continue to improve the docs, what we have in place in this PR is factually accurate and as ready for merge as it can be. Thanks!

**Comment by nate-double-u** [comment]
I'm OK with this merging in as is. We can open new issues to address @sftim's concerns.

/lgtm

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:336]: pls remove line 130-131

**Comment by jimangel** [comment]
After addressing @tengqm's suggestions, I see no issues with this merging, but I differ the final approval to the v1.23 SIG Docs Release Lead @jlbutler 🎉 

It would be great if anyone is willing to open a tracking issue to cover the intended future state, but I wouldn't consider it blocking for _this_ PR.

/lgtm

Thanks @bridgetkromhout!!

**Comment by reylejano** [comment]
Thank you @bridgetkromhout  🚀 
/lgtm

**Comment by bridgetkromhout** [comment]
Thanks, @jimangel & @reylejano! I made the requested change so we do need another LGTM, please and thank you.

> After addressing @tengqm's suggestions, I see no issues with this merging, but I differ the final approval to the v1.23 SIG Docs Release Lead @jlbutler 🎉

Great; as I've addressed @tengqm's note, I'm looking for that approval from @jlbutler! 👋 

> It would be great if anyone is willing to open a tracking issue to cover the intended future state, but I wouldn't consider it blocking for this PR.

I'll do more post-turkey-day. Thanks.

**Comment by jimangel** [comment]
Thanks! 🚀
Would you mind squashing your commits? Happy to reapply a `/lgtm` again after you push. Alternately we can slap a squash label on the PR.
/lgtm

**Comment by bridgetkromhout** [comment]
> Alternately we can slap a squash label on the PR.

@jimangel let's go with this, since it's way too late at night for me to resolve conflicts in the feature-gate list, and I don't want to hold things up. Thanks!

**Comment by sftim** [comment]
@bridgetkromhout if you're happy for me to, I can force-push a squashed version (that would retain LGTM).

**Review by sftim — COMMENTED** [review]
I suggest we also update https://kubernetes.io/docs/tasks/network/validate-dual-stack/ to change the min-kubernetes-server-version, and add a text explanation (ie: you can validate with an earlier version but the feature was only GA and officially supported since v1.23) after the version-check shortcode.

Do we also need to update https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/dual-stack-support/ ?

**Comment by bridgetkromhout** [comment]
@sftim You have the advantage time-zone wise! I'm happy to have you do whatever works best; sounds like we aren't quite ready to squash anyhow, because of these:

> I suggest we also update https://kubernetes.io/docs/tasks/network/validate-dual-stack/ to change the min-kubernetes-server-version, and add a text explanation (ie: you can validate with an earlier version but the feature was only GA and officially supported since v1.23) after the version-check shortcode.

> Do we also need to update https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/dual-stack-support/ ?

I'll look today.

**Comment by sftim** [comment]
Thanks. This PR is in good shape to meet the ready-for-review deadline: it's already there, and now we're looking at snags.

**Comment by bridgetkromhout** [comment]
> Thanks. This PR is in good shape to meet the ready-for-review deadline: it's already there, and now we're looking at snags.

@sftim Thanks! I pushed a PR to address the two things you mentioned today - please take a look - hopefully we can get to squash-and-LGTM-again later today.

**Review by khenidak — APPROVED** [review]

**Review by sftim — COMMENTED** [review]
I spotted one more thing.

We should reword:
> Currently, the kube-controller-manager flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are being left with default values. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).

- remove “Currently”
- reword if technically incorrect for v1.23

Let's address that in a separate PR.

/lgtm

**Comment by sftim** [comment]
@khenidak if _you_ re-add LGTM, I can also approve this (I'd rather not both approve and LGTM commits that _I_ pushed).

**Comment by sftim** [comment]
Also relevant to:
/sig cluster-lifecycle

**Comment by jlbutler** [comment]
I've been monitoring this and looks ok to approve. But looks like you're asking for additional review so you can approve @sftim. Ok by me, just let me know if you'd rather I just merge it. Thanks.

**Comment by jlbutler** [comment]
Chatted with TIm on Slack, thanks for everyone's work on this!

/approve

**Pull request merged**
